### PR TITLE
Make credentials caching  thread safe

### PR DESCRIPTION
**Pull request by pdunna2**
Jira: DISC-593
Reviewer: wkao, ilia

### CODE REVIEW

File: common.go [hunk 1]
```diff
@@ -64,52 +65,86 @@ func serviceAndRegion(host string) (service string, region string) {
...
+
+func (cs *CredentialsStore) getCredentials() (bool, Credentials) {
+	cs.RLock()
+	defer cs.RUnlock()
+
+	if cs.credentials == nil {
+		return false, Credentials{}
+	}
+
+	if !cs.credentials.expired() {
```
Comment (wkao-ellation) [inline common.go]: should we be returning invalid credentials if they have expire instead of not?

File: common.go [hunk 2]
```diff
@@ -64,52 +65,86 @@ func serviceAndRegion(host string) (service string, region string) {
...
+
+func (cs *CredentialsStore) getNewCredentials() Credentials {
+	cs.Lock()
+	defer cs.Unlock()
+
+	if cs.credentials == nil {
+		cs.credentials = getNewKeys()
+	}
+
+	if cs.credentials.expired() {
```
Comment (wkao-ellation) [inline common.go]: nitpick: you can put this as an or statement above

File: common.go (hunk 1 above)
Comment (pdunna2) [reply common.go]: yup , I'll fix it. thnx

**Comment by pdunna2** [comment]
since we only have to protect the credentials pointer ,  it may be possible to use atom.StorePointer and avoid all this locking code.  If somebody has any suggestions on how to do that way , pls le me know

**Comment by iliacimpoes** [comment]
@pdunna2 I think lock is ok for this case, `atomic.Value` is a bit tricky to use

**Comment by iliacimpoes** [comment]
I think we should consider switching to official [aws-sdk-go](https://github.com/aws/aws-sdk-go). It has [sign functionality](https://github.com/aws/aws-sdk-go/blob/master/aws/signer/v4/v4.go), [credentials handling](https://github.com/aws/aws-sdk-go/blob/master/aws/credentials/credentials.go) with [a lot of credentials providers](https://github.com/aws/aws-sdk-go/tree/master/aws/credentials)

File: common.go [hunk 3]
```diff
@@ -64,52 +65,86 @@ func serviceAndRegion(host string) (service string, region string) {
 	return
 }
 
-var credentials *Credentials
-var renewCreds bool
+type CredentialsStore struct {
```
[the remark this answers is not in the document]
Comment (pdunna2) [reply common.go]: @iliacimpoes  In your suggestion you are dereferencing cs.credentials right after Unlock. The moment you dereference a pointer outside a critical section, it is no more thread safe.

+		cs.RUnlock()
+		return *cs.credentials

[the remark this answers is not in the document]
Comment (iliacimpoes) [reply common.go]: yeah, my bad this should like this:
```go
	cs.RLock()
	if cs.credentials != nil && !cs.credentials.expired() {
		credentials := *cs.credentials
		cs.RUnlock()
		return credentials
	}
```

File: common.go [hunk 4]
```diff
@@ -64,52 +65,86 @@ func serviceAndRegion(host string) (service string, region string) {
 	return
 }
```
[the remark this answers is not in the document]
Comment (pdunna2) [reply common.go]: @iliacimpoes Good find,  Please feel free to update the PR with your changes.

**Pull request merged**